### ISSUE TITLE
ElidedSemicolonAndRightBrace in compile error message

### ISSUE THREAD

**Issue by trancexpress** [eclipse-jdt/eclipse.jdt.core#903]
Found during experiments for #859.

To reproduce, compile snippet:

```
public class X {
	public interface S<T> {
	    T s(T t);
	}
    public void f() {
    	S<?> s = v -> v ) ; 
    	{} // empty block to force scope handling for the lambda line
    }
}
```

Observe compile error from `ecj` (4.27):

```
----------
1. ERROR in .../X.java (at line 6)
        S<?> s = v -> v ) ;
                        ^
Syntax error on token ")", ElidedSemicolonAndRightBrace expected
----------
1 problem (1 error)
```

Compile error from `javac` (OpenJDK 17):

```
X.java:6: error: ';' expected
        S<?> s = v -> v ) ;
                       ^
1 error
```

This seems to not be expected (unsurprisingly), see `ProblemReporter.replaceIfSynthetic(String)`:

```
private String replaceIfSynthetic(String token) {
	/* Java 8 grammar changes use some synthetic tokens to make the grammar LALR(1). These tokens should not be exposed in messages
	   as it would make no sense to the programmer whatsoever. Replace such artificial tokens with some "suitable"  alternative. At
	   the moment, there are two synthetic tokens that need such massaging viz : "BeginLambda" and "BeginTypeArguments". There is a
	   third synthetic token "ElidedSemicolonAndRightBrace" that we don't expect to show up in messages since it is manufactured by
	   the parser automatically.
	*/
```

**Comment by mpalat** [comment]
@trancexpress Am planning to work with a couple of newcomers (already identified) for this - hence self-assigned for now.

**Comment by trancexpress** [comment]
> @trancexpress Am planning to work with a couple of newcomers (already identified) for this - hence self-assigned for now.

Sounds great! If possible, can you look into #859, #367, and #125  as well? They seems to be very related to this bug. Possibly all can be fixed with the same change. See also the comment from Stephan here: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/125#issuecomment-1489366453

There seems to be some complicated workaround for lambdas, that is maybe not necessary anymore. To my understanding, this workaround is the cause for the unexpected error message here.

**Comment by srikanth-sankaran** [comment]
I don't think this qualifies as a good-first-issue. Nothing to do with DiagnoseParse qualifies as a "good" issue or a "first" issue let alone a "good first" issue (guffaw)

**Comment by srikanth-sankaran** [comment]
You may want to look at the decidedly low-tech solution adopted for https://github.com/eclipse-jdt/eclipse.jdt.core/issues/367

**Comment by mpalat** [comment]
> I don't think this qualifies as a good-first-issue. Nothing to do with DiagnoseParse qualifies as a "good" issue or a "first" issue let alone a "good first" issue (guffaw)

point taken - removed the assignment for now.

**Comment by srikanth-sankaran** [comment]
The evolving PR https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4667 addresses this issue. We get a diagnostic `Syntax error on token ")", delete this token` which is very precise.
